src/autoscaler.py

```diff
@@ -1,54 +1,99 @@
 import uuid
 from dataclasses import dataclass, field
 import logging
+from ops.pebble import ExecError
 from pathlib import Path
 from typing import Dict
 import yaml
 
 from errors import JujuEnvironmentError
 
 logger = logging.getLogger(__name__)
+CONTROLLER = "scaler-controller"
 
 
 @dataclass
 class AutoScaler:
     controller_data: Dict[str, str] = field(default_factory=dict)
+    model_data: Dict[str, str] = field(default_factory=dict)
     secrets: Dict[str, str] = field(default_factory=dict)
     command: str = ""
 
-    def _build_command(self, juju_config, charm):
-        model, scale = juju_config["default_model_uuid"], juju_config["scale"]
+    def _build_command(self, config, charm):
+        model, scale = config["default_model_uuid"], config["scale"]
         node_groups = scale.nodes(default_model=model.cfg)
         if not node_groups:
             logger.info("Missing juju-scale config")
             raise JujuEnvironmentError("Waiting for Juju Configuration")
 
-        provider_args = "--cloud-provider=juju"
-        nodes_args = " ".join([f"--nodes {node}" for node in node_groups])
-        namespace_args = f"--namespace {charm.model.name.strip()}"
+        namespace = f"--namespace {charm.model.name.strip()}"
+        provider = "--cloud-provider=juju"
+        nodes = " ".join([f"--nodes {node}" for node in node_groups])
 
-        self.command = f"{self.binary} {provider_args} {nodes_args} {namespace_args}"
+        extra_args = config["extra_args"].key_values
+        extra = ""
+        if extra_args:
+            extra = " " + " ".join(
+                sorted(f"--{key}='{value}'" for key, value in extra_args.items())
+            )
+
+        self.command = f"{self.binary} {namespace} {provider} {nodes}{extra}"
```

Comment:
Is this a shell command? Does it need to be escaped for security reasons?

Reply:
I see this gets passed through to the pebble layer via the `command` field. Per the [layer specification](https://github.com/canonical/pebble#layer-specification) this is NOT a shell command, so, this should be safe.